hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java

```diff
@@ -183,53 +206,54 @@ static <T> T tryEachDatanode(Pipeline pipeline,
    */
   public static GetBlockResponseProto getBlock(XceiverClientSpi xceiverClient,
-      List<Validator> validators,
-      DatanodeBlockID datanodeBlockID,
-      Token<? extends TokenIdentifier> token) throws IOException {
-    GetBlockRequestProto.Builder readBlockRequest = GetBlockRequestProto
-        .newBuilder()
-        .setBlockID(datanodeBlockID);
-    ContainerCommandRequestProto.Builder builder = ContainerCommandRequestProto
-        .newBuilder()
+      List<Validator> validators, BlockID blockID, Token<? extends TokenIdentifier> token,
+      Map<DatanodeDetails, Integer> replicaIndexes) throws IOException {
+    ContainerCommandRequestProto.Builder builder = getContainerCommandRequestProtoBuilder()
         .setCmdType(Type.GetBlock)
-        .setContainerID(datanodeBlockID.getContainerID())
-        .setGetBlock(readBlockRequest);
+        .setContainerID(blockID.getContainerID());
     if (token != null) {
       builder.setEncodedToken(token.encodeToUrlString());
     }
 
     return tryEachDatanode(xceiverClient.getPipeline(),
-        d -> getBlock(xceiverClient, validators, builder, d),
-        d -> toErrorMessage(datanodeBlockID, d));
+        d -> getBlock(xceiverClient, validators, builder, blockID, d, replicaIndexes),
+        d -> toErrorMessage(blockID, d));
   }
 
-  static String toErrorMessage(DatanodeBlockID blockId, DatanodeDetails d) {
+  static String toErrorMessage(BlockID blockId, DatanodeDetails d) {
     return String.format("Failed to get block #%s in container #%s from %s",
         blockId.getLocalID(), blockId.getContainerID(), d);
   }
 
   public static GetBlockResponseProto getBlock(XceiverClientSpi xceiverClient,
-      DatanodeBlockID datanodeBlockID,
-      Token<? extends TokenIdentifier> token) throws IOException {
-    return getBlock(xceiverClient, getValidatorList(), datanodeBlockID, token);
+      BlockID datanodeBlockID, Token<? extends TokenIdentifier> token,
+      Map<DatanodeDetails, Integer> replicaIndexes) throws IOException {
+    return getBlock(xceiverClient, getValidatorList(), datanodeBlockID, token, replicaIndexes);
   }
 
-  private static GetBlockResponseProto getBlock(XceiverClientSpi xceiverClient,
-      List<Validator> validators,
-      ContainerCommandRequestProto.Builder builder,
-      DatanodeDetails datanode) throws IOException {
+  private static GetBlockResponseProto getBlock(XceiverClientSpi xceiverClient, List<Validator> validators,
+      ContainerCommandRequestProto.Builder builder, BlockID blockID, DatanodeDetails datanode,
+      Map<DatanodeDetails, Integer> replicaIndexes) throws IOException {
     String traceId = TracingUtil.exportCurrentSpan();
     if (traceId != null) {
       builder.setTraceID(traceId);
     }
+    final DatanodeBlockID.Builder datanodeBlockID = blockID.getDatanodeBlockIDProtobufBuilder();
+    int replicaIndex = replicaIndexes.getOrDefault(datanode, 0);
```

Comment:
Can we get the replica index from the pipeline here, instead of requiring a map to be passed?

```suggestion
    int replicaIndex = xceiverClient.getPipeline().getReplicaIndex(datanode);
```

Reply:
No we cannot use this. My previous commit had the same thing. XceiverClient maintains a cache for the pipeline. 2 pipelines are same if the list of nodes are the same. https://github.com/apache/ozone/blob/8db644ca4643bd4d22f6fe501c62141b8dd4acbc/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientManager.java#L237

Reply:
So 2 GetBlock requests to the same pipeline can interfere with one another. This was the reason why acceptance test was failing previously.

Reply:
I see, so we need to use the refreshed pipeline because indexes may be different.  In that case, I think we can still pass the pipeline instead of the `replicaIndexes` map.

Reply:
ok

Reply:
> I see, so we need to use the refreshed pipeline because indexes may be different. In that case, I think we can still pass the pipeline instead of the `replicaIndexes` map.

It has got nothing to do with the refreshed pipeline. Consider the case a 2 containers having the same nodes but different replica index ordering. So if a key points has blocks belonging to both the containers, so it would reuse the same xceiverClient using having the same pipeline object. This would lead to a getBlock failure for the second block.

Reply:
>2 containers having the same nodes but different replica index ordering
This should not be possible right? Placement policy should not allow this to place 2 different indexes in same node.Do we have a scenario?

Does this impact performance? for every getBlock we refreshPipeline ? Even though client has old index and if it is able to read same index block, then things should be fine right?

Reply:
I had a chat with @swamirishi offline. Current pipeline ids created per ECBlockOutStream is based on dn uuid. There is a chance two distinct containers can be part os same node and pipeline cache can get connection correctly but client also holds pipeline object, which can be pointing to other container rep index as part of dn->replIndex map. 

@swamirishi please provide the description what we discussed here for better understanding. Thanks

Reply:
The standalone pipeline getting initialized sets the pipeline id as the Datanode UUID https://github.com/apache/ozone/blob/c4dc6a072f1171a9a2962ee9a81b91b9d7a6d0ba/hadoop-hdds/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockInputStream.java#L179. Now when a client reads a key, it could be reading the key from multiple blocks belonging to different containers. 
Consider the case:
DN1-uuid = 1 = Standalone PipelineId
Container 1(Block1),    Container 2(Block2)
    DN1 => R3       		DN1 => R2  
The XceiverClientManager maitains a cache based on the pipeline id. So for a standalone pipeline it would be nothing but the DN uuid itself. If the client for the DN has been initialized before then the same client would be reused. But the replicaIndexes for the pipelines are different. Now that we have added a validation on the replication index on the getBlock request. The get Block2 call to getting DN1 will fail if we depend on the pipeline object inside the [xceiverClient object](https://github.com/apache/ozone/blob/2d09598dd352a81fb6a59189b3fe9c6027bdf785/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientSpi.java#L103), since the pipeline object expects a replicaIndex 3 in DN1 but the actual block2 data present in the datanode is for replica index 2. So the server throws an exception because of this.

Reply:
The same could also occur when a refresh pipeline as well, wherein the DNs have changed the replica index, the client has updated the pipeline object but the cached XceiverClient has not updated the replicaIndexes. This is the same case @adoroszlai previously mentioned in the thread.